text/0000-wasm-smart-contracts.md

```diff
@@ -0,0 +1,785 @@
+  - Contract: the first 64 characters of the hash of "name,version"
+
+For example, the address for a contract with name "example" and version "1.0"
+address would be:
+
+```pycon
+  >>> '00ec02' + get_hash("example,1.0")
+  '00ec0248a8e00e3fbca83815668ec5eee730023e6eb61b03b54e8cae1729bf5a0bec64'
+```
+
+
+## Transaction Payload and Execution
+
+Below, the different payload actions are defined along with the inputs and
+outputs that are required in the transaction header.
+
+### SabrePayload
+
+A SabrePayload contains an action enum and the associated action payload. This
+allows for the action payload to be dispatched to the appropriate logic.
+
+Only the defined actions are available and only one action payload should be
+defined in the SabrePayload.
+
+```protobuf
+  message SabrePayload {
+    enum Action {
+      ACTION_UNSET = 0;
+      CREATE_CONTRACT = 1;
+      DELETE_CONTRACT = 2;
+      EXECUTE_CONTRACT = 3;
+      CREATE_CONTRACT_REGISTRY = 4;
+      DELETE_CONTRACT_REGISTRY = 5;
+      UPDATE_CONTRACT_REGISTRY_OWNERS = 6;
+      CREATE_NAMESPACE_REGISTRY = 7;
+      DELETE_NAMESPACE_REGISTRY = 8;
+      UPDATE_NAMESPACE_REGISTRY_OWNERS = 9;
+      CREATE_NAMESPACE_REGISTRY_PERMISSION = 10;
+      DELETE_NAMESPACE_REGISTRY_PERMISSION = 11;
+    }
+
+    Action action = 1;
+
+    CreateContractAction create_contract = 2;
+    DeleteContractAction delete_contract = 3;
+    ExecuteContractAction execute_contract = 4;
+
+    CreateContractRegistryAction create_contract_registry = 5;
+    DeleteContractRegistryAction delete_contract_registry = 6;
+    UpdateContractRegistryOwnersAction update_contract_registry_owners = 7;
+
+    CreateNamespaceRegistryAction create_namespace_registry = 8;
+    DeleteNamespaceRegistryAction delete_namespace_registry = 9;
+    UpdateNamespaceRegistryOwnersAction update_namespace_registry_owners = 10;
+    CreateNamespaceRegistryPermissionAction create_namespace_registry_permission = 11;
+    DeleteNamespaceRegistryPermissionAction delete_namespace_registry_permission = 12;
+  }
+```
+
+### CreateContractAction
+
+Creates a contract and updates the associated contract registry.
+
+```protobuf
+  message CreateContractAction {
+    string name = 1;
+    string version = 2;
+    repeated string inputs = 3;
+    repeated string outputs = 4;
+    bytes contract = 5;
+  }
+```
+
+If a contract with the name and version already exists the transaction is
+considered invalid.
+
+The contract registry is fetched from state and the transaction signer is
+checked against the owners. If the signer is not an owner, the transaction is
+considered invalid.
+
+If the contract registry for the contract name does not exist, the transaction
+is invalid.
+
+Both the new contract and the updated contract registry are set in state.
+
+The inputs for CreateContractAction must include:
+
+* the address for the new contract
+* the address for the contract registry
```

Comment:
Why is the address for the contract registry required? Is that implicit from the contract name? 
If the address is to be used to check validity, then the client shouldn't be allowed to submit it, the TP code should just compute it as ContractRegistry+hash(contract).